File: ExecutionContext/MultipleOrderedEC-Python/MultipleOrderedEC_Lib/MPComp.py
# -*- coding: utf-8 -*-

##
#   @file MPComp.py
#   @brief 実行順序設定関連のクラス、関数

import struct
import OpenRTM_aist
import os.path
import sys
import imp




##
# @class add_Rule
# @brief 実行条件のクラス
#
class add_Rule:
    ##
    # @brief コンストラクタ
    # @param self 
    def __init__(self):
        self.state = -1
        self.name = ""
        self.r = None

##
# @class sub_Rule
# @brief ブロックのクラス
#
class sub_Rule:
    ##
    # @brief コンストラクタ
    # @param self 
    def __init__(self):
        self.v = ""
        self.r = None
        self.s = 0

##
# @class Rule
# @brief 並列ブロックのクラス
#
class Rule:
    ##
    # @brief コンストラクタ
    # @param self 
    def __init__(self):
        self.SR = []

##
# @class main_Rule
# @brief 実行順序のクラス
#
class main_Rule:
    ##
    # @brief コンストラクタ
    # @param self 
    def __init__(self):
        self.rs = []
        self.ar = []

##
# @brief バイナリファイルより文字読み込みする関数
# @param ifs ファイルストリーム
# @return 読み込んだ文字列
def ReadString(ifs):
    s = struct.unpack("i",ifs.read(4))[0]
    a = ifs.read(s)

    return a

##
# @brief バイナリファイルに文字保存する関数
# @param a 書き込む文字列
# @param ofs ファイルストリーム
def WriteString(a, ofs):
    
    a2 = a + "\0"
    s = len(a2)
    
    d = struct.pack("i", s)
    ofs.write(d)
    
    ofs.write(a2)

##
# @brief テキストファイルからの実行順序(直列ブロックのリスト)読み込む関数
# @param cs ファイルから読み込んだ単語のリスト
# @param nm 現在の位置
# @param r 並列ブロック
def LoadSRule(cs, nm, r):
    
    flag = True
    while(flag):

        SR = []

    
        nm = LoadSubRule(cs, nm, SR)

	

	

        r.SR.append(SR)


        nm = nm + 1



    
        if cs[nm] == "}":
            flag = False
            return nm


##
# @brief テキストファイルからの実行順序(並列ブロック)読み込む関数
# @param cs ファイルから読み込んだ単語のリスト
# @param nm 現在の位置
# @param rs 実行順序
def LoadHRule(cs, nm, rs):
    flag = True;
    while(flag):
        r = Rule()

        while(True):

            if cs[nm] == "{":
                break
            elif(cs[nm] == "}"):
                return nm
            else:
                nm = nm + 1

	    


        nm = nm + 1


        nm = LoadSRule(cs,nm,r)
        rs.append(r)

        nm = nm + 1

	

        if cs[nm] == "}":
            flag = False;
            return nm




##
# @brief テキストファイルからの実行順序(実行条件)読み込む関数
# @param cs ファイルから読み込んだ単語のリスト
# @param nm 現在の位置
# @param ar 条件のリスト
def LoadAddRule(cs, nm, ar):
	
	flag = True;
	while(flag):
		ae = add_Rule()
		ae.name = cs[nm]

		nm = nm + 1

	
		
		if cs[nm] == "なし":
			ae.state = -1
		elif cs[nm] == "CREATED":
			ae.state = OpenRTM_aist.RTC.CREATED_STATE
		elif cs[nm] == "INACTIVE":
			ae.state = OpenRTM_aist.RTC.INACTIVE_STATE
		elif cs[nm] == "ACTIVE":
			ae.state = OpenRTM_aist.RTC.ACTIVE_STATE
		elif cs[nm] == "ERROR":
			ae.state = OpenRTM_aist.RTC.ERROR_STATE
		
		
	
	
		ar.append(ae)

		nm = nm + 1

        

		
		if cs[nm] == ";":
			flag = False
			return nm
		elif cs[nm] == "}":
			flag = False
			return nm
		
		

##
# @brief テキストファイルからの実行順序(直列ブロック)読み込む関数
# @param cs ファイルから読み込んだ単語のリスト
# @param nm 現在の位置
# @param sr 条件のリスト
def LoadSubRule(cs, nm, sr):
    flag = True
    while(flag):
        rs = sub_Rule()

        rs.v = cs[nm]

        sr.append(rs)


        nm = nm + 1

	


    
        if cs[nm] == ";":
            flag = False
            return nm
        elif cs[nm] == "}":
            flag = False;
            return nm


##
# @brief ファイルより実行順序の読み込む関数
# @param rs 実行順序
# @param fName ファイル名
# @return 成功でTrue、失敗でFalse
def LoadMainRule(rs, fName):
    
    if LoadMainRule == "":
        R = main_Rule()
        rs.append(R)
        return True

    root, ext = os.path.splitext(fName)
    

    if ext == ".py":
        try:
            pathName = os.path.dirname(fName)
            fileName = os.path.basename(root)
            
            (file, pathname, description) = imp.find_module(fileName,[pathName])
            plugin = imp.load_module(fileName, file, pathname, description)
            Ls = plugin.CompList
            for h in range(0, len(Ls)):
                R = main_Rule()
                for i in range(0, len(Ls[h][0])):
                    ar = add_Rule()
                    ar.name = Ls[h][0][i][0]
                    ar.state = Ls[h][0][i][1]
                    
                    R.ar.append(ar)
                for i in range(0, len(Ls[h][1])):
                    r = Rule()
                    
                    for j in range(0, len(Ls[h][1][i])):
                        
                        SRs = []
                        for k in range(0, len(Ls[h][1][i][j])):
                            SR = sub_Rule()
                            SR.v = Ls[h][1][i][j][k]
                            SR.s = 0
                            SRs.append(SR)
                        r.SR.append(SRs)
                    R.rs.append(r)
                rs.append(R)
        except:
            R = main_Rule()
            rs.append(R)
            return False


    else:
        try:
            infile = open(fName, 'rb')
        except:
            R = main_Rule()
            rs.append(R)
            return False
        
        count = 0

        Type = 0
        while True:
            count += 1
            c = infile.read(1)
            if c =="":
                break
            c = struct.unpack("b",c)[0]
            if c == 0x00:
                Type = 1
                
        infile.seek(0)

        if count < 2:
            R = main_Rule()
            rs.append(R)
            infile.close()
            return False





        if Type == 0:
            
            try:
                infile = open(fName)
            except:
                R = main_Rule()
                rs.append(R)
                return False
            line = infile.readline()

            cs = []

            while line:
                lines1 = line
                lines1 = lines1.replace("\t","")
                lines1 = lines1.replace("\n","")
                lines1 = lines1.split(' ')
                for d in lines1:
                    if d != "":
                        cs.append(d)
                line = infile.readline()
            infile.close

            nm = 0
            while(len(cs) > nm):
                if cs[nm] == "{":
                    mr = main_Rule()

                    while(True):
                        nm = nm + 1
                        if cs[nm] == "{":
                            break
                    nm = nm + 1

            
                                    
                    nm = LoadAddRule(cs, nm, mr.ar)
            

            
                                    

                    while True:
                
                        nm = nm + 1
                        
                        if cs[nm] == "{":
                            break
            
            
                    nm = nm + 1



                    nm = LoadHRule(cs, nm, mr.rs)

            

                    while True:
                        nm = nm + 1
                            
                        if cs[nm] == "}":
                            break
            
                    nm = nm + 1
                    rs.append(mr)
        else:
            m = struct.unpack("i",infile.read(4))[0]
            for h in range(0, m):
                R = main_Rule()
                w = struct.unpack("i",infile.read(4))[0]
                for i in range(0, w):
                    ar = add_Rule()
                    ar.name = ReadString(infile)
                    ar.name = ar.name.replace("\0","")
                    

                    a_t = struct.unpack("i",infile.read(4))[0]

                    
                    
                    a_v = -1
                    if a_t == -1:
                        a_v = -1
                    elif a_t == 0:
                        a_v = OpenRTM_aist.RTC.CREATED_STATE
                    elif a_t == 1:
                        a_v = OpenRTM_aist.RTC.INACTIVE_STATE
                    elif a_t == 2:
                        a_v = OpenRTM_aist.RTC.ACTIVE_STATE
                    elif a_t == 3:
                        a_v = OpenRTM_aist.RTC.ERROR_STATE
                        
                    ar.state = a_v

                    R.ar.append(ar)
                c = struct.unpack("i",infile.read(4))[0]
                
                for i in range(0, c):
                    r = Rule()
                    d = struct.unpack("i",infile.read(4))[0]

                    

                    for j in range(0, d):
                        e = struct.unpack("i",infile.read(4))[0]
                        
                        SRs = []
                        

                        for k in range(0, e):
                            
                            SR = sub_Rule()
                            SR.v = ReadString(infile)
                            
                            SR.v = SR.v.replace("\0","")
                            
                            SR.s = 0
                            SRs.append(SR)
                        r.SR.append(SRs)
                    R.rs.append(r)
                rs.append(R)
                

                     




        infile.close()

    return True

File: ExecutionContext/MultipleOrderedEC-C++/MPComp.cpp
/*!
* @file  MPComp.cpp
* @brief 実行順序設定関連のクラス、関数
*
*/

#include <coil/stringutil.h>
#include "MPComp.h"
#include "FileStreamFunc.h"
#include "MPLua.h"

using namespace std;


/**
*@brief 直列ブロック追加
*/
void Rule::addP()
{
	std::vector<sub_Rule> s;
	SR.push_back(s);
}

/**
*@brief 指定番号の直列ブロックにブロック追加
* @param num 番号
* @param s ブロックのRTC名
*/
void Rule::addsubRule(int num, std::string v)
{
	if(SR.size() > num)
	{
		sub_Rule s;
		s.v = v;
		s.s = 0;
		SR[num].push_back(s);
	}
}

/**
*@brief 並列ブロック追加
*/
void main_Rule::addRule()
{
	Rule r;
	rs.push_back(r);
}

/**
*@brief 実行条件追加
* @param n RTC名
* @param s 状態
*/
void main_Rule::addaddRule(std::string n, int s)
{
	add_Rule a;
	a.name = n;
	a.state = s;
	ar.push_back(a);
}

/**
*@brief 指定した番号の並列ブロックを取得
* @param num 番号
* @return 並列ブロック
*/
Rule *main_Rule::getRule(int num)
{
	if(rs.size() > num)
		return &rs[num];
	return NULL;
}

/**
*@brief 指定した番号の実行条件を取得
* @param num 番号
* @return 実行条件
*/
add_Rule *main_Rule::getaddRule(int num)
{
	if(ar.size() > num)
		return &ar[num];
	return NULL;
}




/**
*@brief テキストファイルからの実行順序(直列ブロックのリスト)読み込む関数
* @param cs ファイルから読み込んだ単語のリスト
* @param nm 現在の位置
* @param r 並列ブロック
*/
void LoadSRule(std::vector<std::string> &cs, int &nm,Rule &r)
{
	bool flag = true;
	while(flag)
	{
		
		std::vector<sub_Rule> SR;
		
		LoadSubRule(cs, nm, SR);

		r.SR.push_back(SR);

		
		if(!AddCount(cs, nm))
			return;
		else if(cs[nm] == "}")
		{
			flag = false;
			return;
		}
		
	}
}


/**
*@brief テキストファイルからの実行順序(並列ブロック)読み込む関数
* @param cs ファイルから読み込んだ単語のリスト
* @param nm 現在の位置
* @param rs 実行順序
*/
void LoadHRule(std::vector<std::string> &cs, int &nm,std::vector<Rule> &rs)
{
	bool flag = true;
	while(flag)
	{
		Rule r;
		
		
		

		while(true)
		{
			
			if(cs[nm] == "{")
				break;
			else if(cs[nm] == "}")
				break;
			else if(!AddCount(cs, nm))
			{
				flag = false;
				return;
			}
		}


		if(!AddCount(cs, nm))
			return;
		

		LoadSRule(cs,nm,r);
		rs.push_back(r);

		if(!AddCount(cs, nm))
			return;
		else if(cs[nm] == "}")
		{
			flag = false;
			return;
		}
		
		
	}
}

/**
*@brief 単語のリストの位置を移動する
* @param cs ファイルから読み込んだ単語のリスト
* @param nm 現在の位置
* @return 成功でTrue、失敗でFalse
*/
bool AddCount(std::vector<std::string> &cs, int &nm)
{
	nm++;
	if(cs.size() <= nm)
	{
		return false;
	}
	return true;
}

/**
*@brief テキストファイルからの実行順序(実行条件)読み込む関数
* @param cs ファイルから読み込んだ単語のリスト
* @param nm 現在の位置
* @param ar 条件のリスト
*/
void LoadAddRule(std::vector<std::string> &cs, int &nm,std::vector<add_Rule> &ar)
{
	

	bool flag = true;
	while(flag)
	{
		add_Rule ae;
		ae.name = cs[nm];
		
		if(!AddCount(cs, nm))
			return;
		if(cs[nm] == "なし"){
			ae.state = -1;
		}
		else if(cs[nm] == "CREATED"){
			ae.state = RTC::CREATED_STATE;
		}
		else if(cs[nm] == "INACTIVE"){
			ae.state = RTC::INACTIVE_STATE;
		}
		else if(cs[nm] == "ACTIVE"){
			ae.state = RTC::ACTIVE_STATE;
		}
		else if(cs[nm] == "ERROR"){
			ae.state = RTC::ERROR_STATE;
		}
		
		
		
		ar.push_back(ae);

		if(!AddCount(cs, nm))
			return;
		if(cs[nm] == ";")
		{
			flag = false;
			return;
		}
		else if(cs[nm] == "}")
		{
			flag = false;
			return;
		}
		
		
	}
}

/**
*@brief テキストファイルからの実行順序(直列ブロック)読み込む関数
* @param cs ファイルから読み込んだ単語のリスト
* @param nm 現在の位置
* @param sr 条件のリスト
*/
void LoadSubRule(std::vector<std::string> &cs, int &nm,std::vector<sub_Rule> &sr)
{
	bool flag = true;
	while(flag)
	{
		sub_Rule rs;
		
		rs.v = cs[nm];
		

		
		
		sr.push_back(rs);

		


		if(!AddCount(cs, nm))
			return;
		else if(cs[nm] == ";")
		{
			flag = false;
			return;
		}
		else if(cs[nm] == "}")
		{
			flag = false;
			return;
		}
		
		
	}
}

/**
*@brief ファイルより実行順序の読み込む関数
* @param rs 実行順序
* @param Name ファイル名
* @return 成功でTrue、失敗でFalse
*/
bool LoadMainRule(std::vector<main_Rule> &rs, std::string Name)
{
	if (Name == "")
	{
		main_Rule mr;
		rs.push_back(mr);

		return true;
	}
#ifdef _WINDOWS
	coil::replaceString(Name, "/", "\\");
#else
	coil::replaceString(Name, "\\", "/");
#endif
	PathList pl = PathList(Name);
	
	
	if(pl.cxt == "lua")
	{
		return LoadMainRule_Lua(rs, Name);
	}
	else
	{
		
		std::ifstream ifs( Name.c_str() , ios::in | ios::binary );
		if (!ifs)
		{
			main_Rule mr;
			rs.push_back(mr);

			return false;
		}

		int count = 0;
		int type = 0;
		while(!ifs.eof())
		{  
			count += 1;
			char c;
			ifs.read( ( char * ) &c, sizeof( char ) );
			if(c == 0)
			{
				type = 1;
			}
			
		}

		
		//ifs.seekg(0, ios_base::beg);
		//ifs.seekg(0, ios_base::beg);

		ifs.close();

		//std::cout << count << std::endl;
		if (count < 2)
		{
			main_Rule mr;
			rs.push_back(mr);

			return true;

		}
		

		if(type == 0)
		{
			
			
			ifs.open(Name.c_str());

			std::vector<std::string> cs;

			string hoge;

			while(getline( ifs, hoge ))
			{
				

				string token;
				istringstream stream( hoge );

				while(getline( stream, token, ' ' ))
				{
					std::string tmp = token;
					
					coil::replaceString(tmp,"	", "");
					coil::replaceString(tmp," ", "");
					
					cs.push_back(tmp);
					//std::cout << cs[cs.size()-1] << std::endl;
					
				}
			}

			/*for(int i=0;i < cs.size();i++)
			{
				
			}*/

			

			for(int i=0;i < cs.size();i++)
			{
				
				if(cs[i] == "{")
				{
					
					
					main_Rule mr;

					while(true)
					{
						if(!AddCount(cs, i))
							break;
						if(cs[i] == "{")
							break;
					}
					
					if (!AddCount(cs, i))
					{
						rs.push_back(mr);
						ifs.close();
						return false;
					}
						

					
					LoadAddRule(cs, i, mr.ar);

					
					

					while(true)
					{
						if(!AddCount(cs, i))
							break;
						if(cs[i] == "{")
							break;
					}
					if(!AddCount(cs, i))
						break;



					LoadHRule(cs, i, mr.rs);

					while(true)
					{
						if(!AddCount(cs, i))
							break;
						if(cs[i] == "}")
							break;
					}
					
					rs.push_back(mr);

					
				}
			}
				
		}
		else
		{
			ifs.open( Name.c_str() , ios::in | ios::binary );

			int m;
			ifs.read( (char*)&m, sizeof(m) );
			
			
			
			for(int h=0;h < m;h++)
			{
				
				main_Rule R;
				int w;
				ifs.read( (char*)&w, sizeof(w) );

				for(int i=0;i < w;i++)
				{
					add_Rule ar;

					string s;
					s = ReadString( ifs );
					ar.name = s;

					int v;
					ifs.read( (char*)&v, sizeof(v) );
					ar.state = v;
					
					
					
					R.ar.push_back(ar);

				}


				int c;
				ifs.read( (char*)&c, sizeof(c) );
				

				for(int i=0;i < c;i++)
				{
					Rule r;
					
					int d;
					ifs.read( (char*)&d, sizeof(d) );

					

					
					for(int j=0;j < d;j++)
					{
						int e;
						ifs.read( (char*)&e, sizeof(e) );

						vector<sub_Rule> SRs;

						

						
						

						for(int k=0;k < e;k++)
						{
							sub_Rule SR;
							string s;
							s = ReadString( ifs );


							SR.v = s;

							SR.s = 0;

							
							SRs.push_back(SR);
						}
						r.SR.push_back(SRs);
					}

					

					R.rs.push_back(r);
				}
				rs.push_back(R);
			}
		}

		ifs.close();
	}

	return true;

	
}


